### PR TITLE
Helper functions 

### PR DESCRIPTION
**Pull request by eraly**
Helper functions
So users can save "featurized" datasets to disk and fit with just the unfrozen graph to iterate quickly. Currently only supporting computation graphs.

### CODE REVIEW

File: deeplearning4j-core/src/test/java/org/deeplearning4j/nn/transferlearning/TransferLearningHelperTest.java [hunk 1]
```diff
@@ -156,6 +156,9 @@ public void testFitUnFrozen() {
         assertEquals(modelIdentical.getLayer("denseLeft0").params(),modelToTune.getLayer("denseLeft0").params());
         assertEquals(modelIdentical.getLayer("outLeft").params(),modelToTune.getLayer("outLeft").params());
 
+        System.out.println(modelIdentical.summary());
```
Comment (agibsonccc) [inline deeplearning4j-core/src/test/java/org/deeplearning4j/nn/transferlearning/TransferLearningHelperTest.java]: Logging here? (log.info)

File: deeplearning4j-core/src/test/java/org/deeplearning4j/nn/transferlearning/TransferLearningComplex.java [hunk 2]
```diff
@@ -160,13 +159,27 @@ public void testWithMergeAndSubset() {
...
             }
             leftGraph.fit(new DataSet(subsetLeft, labels[0]));
             centreGraph.fit(new DataSet(denseCentre2, labels[1]));
+            assertEquals(rightDataSet,new MultiDataSet(new INDArray[] {modelNow.feedForward(features,false).get("denseCentre2"),rightDataSet.getFeatures(1)},new INDArray[] {labels[2]}));
+
             rightGraph.fit(rightDataSet);
             modelNow.fit(new MultiDataSet(features, labels));
+            assertEquals(modelNow.getLayer("denseCentre2").params(),frozenGraph.getLayer("denseCentre2").params());
             System.out.println("Fit after "+n);
             for (int i = 0; i < listOfLayers.length; i++) {
```
Comment (agibsonccc) [inline deeplearning4j-core/src/test/java/org/deeplearning4j/nn/transferlearning/TransferLearningComplex.java]: System.out.println -> log.info

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/transferlearning/TransferLearningHelper.java [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+
+        if (frozenInputVertices.isEmpty()) {
+            throw new IllegalArgumentException("No frozen layers found");
+        }
+
+    }
+
+    /*
+    private void initHelperMLN() {
+
```
Comment (agibsonccc) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/transferlearning/TransferLearningHelper.java]: Needs to be implemented yet?

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/transferlearning/TransferLearningHelper.java [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+        //find a non frozen vertex that has a frozen parent which will always be a layer vertex
+
+        //outputs are the same
+        //remove some input vertices
+        //set new inputs
+
+    }
+    */
+
+    public MultiDataSet featurize(MultiDataSet input) {
```
Comment (agibsonccc) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/transferlearning/TransferLearningHelper.java]: Even stub javadoc comments would be good here so it picks it up when we generate it.

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/transferlearning/TransferLearningHelper.java [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+     */
+    private void initHelperGraph() {
+
+        //parent vertices added in when seen
+        Set<String> seenAsParents = new HashSet<>();
+        int [] backPropOrder = origGraph.topologicalSortOrder().clone();
+        ArrayUtils.reverse(backPropOrder);
+
+        for (int i = 0; i<backPropOrder.length;i++) {
+            GraphVertex currentVertex = origGraph.getVertices()[backPropOrder[i]];
```
Comment (agibsonccc) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/transferlearning/TransferLearningHelper.java]: Spacing.

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/transferlearning/TransferLearningHelper.java [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+            String currentName = currentVertex.getVertexName();
+            if (!currentVertex.hasLayer()) {
+                //before skipping over a subset vertex check if it has a frozen parent
+                if (currentVertex instanceof SubsetVertex) {
+                    //if this is a subset vertex and has a parent that is a frozen layer that has not been seen
+                    //add to list of frozen inputs
+                    VertexIndices[] parentVertices = currentVertex.getInputVertices();
+                    for(int j=0; j<parentVertices.length; j++ ) {
```
Comment (agibsonccc) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/transferlearning/TransferLearningHelper.java]: Spacing.

**Comment by eraly** [comment]
It's a WIP. I have to add java doc and maybe more tests.

File: deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/transferlearning/TransferLearningHelper.java (hunk 3 above)
Comment (eraly) [inline deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/transferlearning/TransferLearningHelper.java]: Yes. The computation graph is done and that is the more complex case. The MLN is pretty straightforward.

**Pull request merged**